data-pipeline-flink/notification-engine/src/main/scala/org/sunbird/dp/notification/function/FirstCourseEnrollmentFunction.scala

```diff
@@ -0,0 +1,239 @@
+      val requestBody = new util.HashMap[String, Any]()
+      requestBody.put(enrollmentConfig.REQUEST, request)
+      val uri=enrollmentConfig.SB_SERVICE_URL+enrollmentConfig.SUNBIRD_USER_SEARCH_ENDPOINT
+      val responseString = restApiUtil.post(uri, requestBody)
+      val gson = new Gson()
+      val responseObj = gson.fromJson(responseString, classOf[util.HashMap[String, Any]])
+      if (responseObj != null && enrollmentConfig.OK.equalsIgnoreCase(responseObj.get(enrollmentConfig.RESPONSE_CODE).toString)) {
+        val result: util.Map[String, Any] = responseObj.get(enrollmentConfig.RESULT).asInstanceOf[util.Map[String, Any]]
+        if (result.get(enrollmentConfig.RESPONSE) != null) {
+          val response: util.Map[String, Any] = result.get(enrollmentConfig.RESPONSE).asInstanceOf[util.Map[String, Any]]
+          val contentList: util.List[util.Map[String, Any]] = response.get(enrollmentConfig.CONTENT).asInstanceOf[util.List[util.Map[String, Any]]]
+          contentList.forEach(content => {
+            if (content.containsKey(enrollmentConfig.PROFILE_DETAILS)) {
+              val profileDetails: util.Map[String, Any] = content.get(enrollmentConfig.PROFILE_DETAILS).asInstanceOf[util.Map[String, Any]]
+              if (profileDetails.containsKey(enrollmentConfig.PERSONAL_DETAILS)) {
+                val personalDetails: util.Map[String, Any] = profileDetails.get(enrollmentConfig.PERSONAL_DETAILS).asInstanceOf[util.Map[String, Any]]
+                if (MapUtils.isNotEmpty(personalDetails)) {
+                  val email = personalDetails.get(enrollmentConfig.PRIMARY_EMAIL).toString
+                  userMap.put(enrollmentConfig.FIRST_NAME,content.get(enrollmentConfig.FIRST_NAME))
+                  userMap.put(enrollmentConfig.LAST_NAME,content.get(enrollmentConfig.LAST_NAME))
+                  userMap.put(enrollmentConfig.PRIMARY_EMAIL,email)
+                }
+              }
+            }
+          })
+        }
+      }
+      userMap
+    }
+    catch {
+      case e: Exception => e.printStackTrace()
+        logger.info(String.format("Failed during fetching user details %s", e.getMessage()))
+        null
+    }
+  }
+
+  def getCourseNameById(courseId: String) :String = {
+    try{
+      var courseName=""
+      val filters = new util.HashMap[String, Any]()
+      filters.put(enrollmentConfig.IDENTIFIER,courseId)
+      filters.put(enrollmentConfig.STATUS,new util.ArrayList[String](){
+        add("Live")
+      })
+      val request = new util.HashMap[String, Any]()
+      request.put(enrollmentConfig.FILTERS, filters)
+      request.put(enrollmentConfig.FIELDS, enrollmentConfig.CONTENT_SEARCH_FIELDS)
+      val requestBody = new util.HashMap[String, Any]()
+      requestBody.put(enrollmentConfig.REQUEST, request)
+      val url: String = enrollmentConfig.KM_BASE_HOST + enrollmentConfig.content_search
+      val obj = restApiUtil.post(url, requestBody)
```

Comment:
There is a mixed use of fetching the user / course details either through the rest API end point or direct from Elastic Search. Can we have some kind of wrapper around it so the user fetching / content fetching can be done through a specific service class and respective transformation can be done inside it. 

This way if there is a logic change to be done while fetching, we can do it in single place rather than doing it in FirstCourseEnrollmentFunction and IncompleteCourseReminderEmailNotification seperately.

Explore on the code re-usability, which will help in maintainability.

Reply:
This thing I guess I addressed for collecting user details now using Elastic search and for getting course details used rest api call . since for getting user details the logic is same in every class for that I created a new class UserUtility and there is previously one class I created for rest call. Also in the enhancement I create one separate class for kafka message intitate